### PR TITLE
Adding Startup Job Runner

### PR DESCRIPTION
**Pull request by gwendolyngoetz**
### Description

Adding Startup Job Runner

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: server/app/durablejobs/DurableJobRegistry.java [hunk 1]
```diff
@@ -72,30 +75,57 @@ public void registerWithNoTimeResolver(
...
+        jobName.getJobNameString(),
+        RegisteredJob.create(
+            durableJobFactory, jobName, jobType, /* jobExecutionTimeResolver= */ Optional.empty()));
+  }
+
+  /** Registers a factory for a given job name. Cannot be used with {$JobType.RECURRING}. */
+  public void registerStartupJob(
+      DurableJobName jobName, JobType jobType, DurableJobFactory durableJobFactory) {
+    validateJobName(jobName);
+    if (jobType == JobType.RECURRING) {
```
Comment (dkatzz) [inline server/app/durablejobs/DurableJobRegistry.java:88]: **suggestion:** Can we use the allowedJobTypes instead here and below? That way we don't need to remember to update this if there becomes more than one recurring job type?

File: server/app/modules/DurableJobModule.java [hunk 2]
```diff
@@ -147,4 +158,23 @@ public DurableJobRegistry provideRecurringDurableJobRegistry(
 
     return durableJobRegistry;
   }
+
+  @Provides
+  @StartupJobsProviderName
+  public DurableJobRegistry provideStartupDurableJobRegistry() {
+    //  PersistedDurableJobRepository persistedDurableJobRepository) {
```
Comment (dkatzz) [inline server/app/modules/DurableJobModule.java:165]: **suggestion:** Can we put a todo here instead with a bug tracking this work?

File: server/test/durablejobs/StartupDurableJobRunnerTest.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    PersistedDurableJobModel jobB = createPersistedJobToExecute();
+    PersistedDurableJobModel jobC = createPersistedJobScheduledInFuture();
+
+    recurringDurableJobRunner.runJobs();
+
+    jobA.refresh();
+    jobB.refresh();
+    jobC.refresh();
+
+    // This assertion fails occasionally. I've been unable to figure out why
```
Comment (dkatzz) [inline server/test/durablejobs/StartupDurableJobRunnerTest.java:133]: **question:** Does this comment still apply?

**Review by dkatzz — APPROVED** [review]

File: server/app/modules/DurableJobModule.java (hunk 2 above)
Comment (gwendolyngoetz) [reply server/app/modules/DurableJobModule.java:165]: Typically I would, but I'm going to be adding the first job in the next day for the service area work I'm doing.

File: server/test/durablejobs/StartupDurableJobRunnerTest.java (hunk 3 above)
Comment (gwendolyngoetz) [reply server/test/durablejobs/StartupDurableJobRunnerTest.java:133]: I don't think so. I haven't had it fail at all.

File: server/app/durablejobs/DurableJobRegistry.java (hunk 1 above)
Comment (gwendolyngoetz) [reply server/app/durablejobs/DurableJobRegistry.java:88]: Not cleanly without a bunch of refactoring. You try to pass in the scheduler you'll get a circular reference. Passing in just the list of allowed types would let the caller pass anything, at which point why ask for allowed types if you can get around the safety check.

Comment (dkatzz) [reply server/app/durablejobs/DurableJobRegistry.java:88]: Ah yeah makes sense - I wonder if its worth a comment above the allowedJobTypes for the recurring class that we'll need to update this if that changes?

**Pull request merged**